### PR TITLE
Replace `snipes` with `cleanByImpersonation` function

### PR DESCRIPTION
**Pull request by espendk**
This PR removes `snipes` and `snipesFor` from Mangrove and replaces them with a `cleanByImpersonation` function that only allows failing offers to be executed.

# Major changes

In addition to the primary change (not allowing snipes in general) the following changes have been made:

## `MgvCleaner` has been removed
Mangrove core now provides the cleaning functionality that `MgvCleaner` used to provide, so it `MgvCleaner` has been removed.

##  Cleaning of multiple offers is more resilient
Before, `MgvCleaner.collect*` would revert if execution of any offer succeeded. The new `cleanByImpersonation` is more resilient as it will not revert the whole tx if an offer succeeds but will revert just the execution of that offer and continue cleaning of the next offers. 

Thus a cleaner that attempts to clean multiple offers in one tx will be able to successfully clean some offers despite some offers failing to be cleaned.

## Taker impersonation no longer checks permits
The new `cleanByImpersonation` function has integrated the `taker` parameter of the old `snipesFor`. This allows cleaning by impersonation, ie where another taker's funds and approval of Mangrove for transfers of quote can be used to execute an offer and make it fail (which was already possible via `MgvCleaner.collectByImpersonation`).

This has been integrated into the `cleanByImpersonation` function instead of having a separate `cleanFor` function because
1. no permit/allowance is needed nor does it have to be checked and
2. cleaners are expected to be advanced users, so requiring an explicit `taker` parameter with non-obvious semantics seems ok.

# Migration of tests away from snipes
Many tests rely on sniping. To allow gradual migration away from this, the snipe functions have been moved to a `TestSnipingMangrove` contract that `MangroveTest` provides to tests.

To make clear that these tests do something non-standard, the snipe functions have been renamed to `snipesInTest` and `snipesForInTest`.

Migration of tests will be done in later PR's as merging this PR will (1) hopefully stop further use of snipes in tests and (2) enable that we can start updating downstream components to the new MGV V2 API.

### CODE REVIEW

File: src/MgvOfferTaking.sol
Comment (adhusson) [inline src/MgvOfferTaking.sol]:
```suggestion
      /* FIXME: edit comment: If we removed the `isLive` conditional, a single expired or nonexistent offer in `targets` would revert the entire transaction (by the division by `offer.gives` below since `offer.gives` would be 0). We also check that `gasreq` is not worse than specified. A taker who does not care about `gasreq` can specify any amount larger than $2^{24}-1$. A mismatched price will be detected by `execute`. */
```

File: src/MgvOfferTakingWithPermit.sol
Comment (adhusson) [inline src/MgvOfferTakingWithPermit.sol]: Removing `deductSenderAllowance` is scary even though I do not see an attack right now.

**Comment by espendk** [comment]
~~Just realized that the old snipe tests that test the details of execution (eg events emitted and transfers made) should be migrated as part of this PR. Will look at that Monday!~~

On second thought (and seeing #444 building on this PR) I think we should merge this after review and include the offer execution tests in the upcoming test migration PR. The feature will still be appropriately tested before merging to `develop`.

File: src/MgvOfferTaking.sol
Comment (lnist) [inline src/MgvOfferTaking.sol]: Do we have a good use case for non-exact match? Simplified code seems like a win.

File: src/MgvOfferTaking.sol
Comment (lnist) [inline src/MgvOfferTaking.sol]: also a comment that needs to be updated. You added a FIXME further up, maybe also here? Though I suspect the code has quite a bit of outdated comments now.

File: src/MgvOfferTaking.sol
Comment (lnist) [inline src/MgvOfferTaking.sol]: Do we leave this FIXME behind?

File: src/MgvOfferTaking.sol
Comment (lnist) [inline src/MgvOfferTaking.sol]: Previously we could clean multiple offers and then in the end do the transfer (at cast 21000 gas). We now pay that amount of gas for each transfer, thus upping the needed penalty to cover the cost.

Should gasbase compensate a normal market order taker, or a cleaner? If it is to incentivize cleaning, then that is the important one.

File: src/MgvOfferTaking.sol
Comment (lnist) [inline src/MgvOfferTaking.sol:410]: This looks correct - however, would it be safer to explicitly check the allowed failure values? 
`require(mgvData == "mgv/makerRevert" || mgvData == "mgv/makerTransferFail" || mgvData == "mgv/makerReceiveFail", "mgv/clean/offerDidNotFail");

File: src/periphery/MgvCleaner.sol
Comment (lnist) [inline src/periphery/MgvCleaner.sol]: takerToImpersonate is now just `taker`, right?

File: test/core/Gatekeeping.t.sol
Comment (lnist) [inline test/core/Gatekeeping.t.sol:661]: is snipesKO replaces by cleanKO?

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol]: I'm wondering about the opposite: should we completely remove the price match? Initially it's there for "normal snipes" (to protect the taker). In the absence of normal snipes, what happens if we just ignore the price?

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol:410]: The intent was to be passively future-proof against additional errors that would be added later. I think that still applies so I would tend towards keeping it as is.

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: All of the comments in that file should be updated :-) I made a general comment on this (and @adhusson added a specific one) and I think it'll be easiest to do it in one go once the `feat/ticks` has stabilized. But I can be persuaded otherwise :-)

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: Good point, I'll remove it.

Comment (adhusson) [reply src/MgvOfferTaking.sol]: To answer the FIXME: `executeEnd` is a noop on Mangrove anyway but is useful on InvertedMangrove (it calls the taker contract).

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol]: I agree -- gasbase should compensate max(market order overhead, cleaning overhead). 

I didn't get the 21k gas reference?

File: src/MgvOfferTakingWithPermit.sol
Comment (adhusson) [reply src/MgvOfferTakingWithPermit.sol]: I would rather keep it in :)

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: Good point!

I think this is a strong reason for moving the multi-clean functionality from `MgvCleaner` to the core. Should be pretty quick to do.

Unless @adhusson objects, I'll go ahead and make the change.

**Review by adhusson — APPROVED** [review]
lgtm pending comments above

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol:410]: Hmm, the point here is to revert explicitly if the offer succeeds, because then it shouldn't be cleaned.
But maybe that isn't clear?

File: src/periphery/MgvCleaner.sol
Comment (espendk) [reply src/periphery/MgvCleaner.sol]: Indeed, I'll fix.

File: test/core/Gatekeeping.t.sol
Comment (espendk) [reply test/core/Gatekeeping.t.sol:661]: Yes, I missed a rename when adapting the reentrancy tests and fixed on the test migration branch - I didn't expect you to notice ;-)

I'll port the fix here!

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: I think the cleaner risks spending more gas on failing to make the offer fail: The offer is now more expensive and no longer fails.

File: src/MgvOfferTakingWithPermit.sol
Comment (adhusson) [reply src/MgvOfferTakingWithPermit.sol]: @espendk convinced me

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: So `executeEnd` should still be invoked when an offer fails, right?

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: The scenario is that the maker updates the offer between us making the decision to clean and actually cleaning - I don't think we can a priori say anything about whether it would still fail. Skipping early saves gas - in case of failure still, we win gas. I don't know what is most common.
I lean towards skipping since information has changed - spending gas on executing the entire offer if it succeeds is quite wasteful.

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: Done in [e4cfb66](https://github.com/mangrovedao/mangrove-core/pull/437/commits/e4cfb668bd6ebfb2547952379be91c8c510a1482)

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: Reintroduce `executeEnd` in [76393f2](https://github.com/mangrovedao/mangrove-core/pull/437/commits/76393f2b9000721d1456705f839d761e0d1e9244)

File: src/MgvOfferTaking.sol
Comment (espendk) [reply src/MgvOfferTaking.sol]: I went for the strict match and `takerWants` instead of `fillVolume` and `fillWants` in [92eff73](https://github.com/mangrovedao/mangrove-core/pull/437/commits/92eff7372aa66d7dd695a6b7a741334d3f9add01)

**Pull request merged**
